### PR TITLE
filter event notifications by restrictions

### PR DESCRIPTION
**Pull request by escattone**
mozilla/sumo#1664

### CODE REVIEW

File: kitsune/kbforums/events.py [hunk 1]
```diff
@@ -103,9 +108,11 @@ def __init__(self, post):
 
     def send_emails(self, exclude=None):
         """Notify watches of the document and of the locale."""
-        return EventUnion(
-            self, NewThreadEvent(self.post), NewThreadInLocaleEvent(self.post)
-        ).send_emails(exclude=exclude)
+        return EventUnion(self, NewThreadInLocaleEvent(self.post)).send_emails(exclude=exclude)
```
Comment (escattone) [inline kitsune/kbforums/events.py:111]: When working on this, I noticed that the second argument to the `EventUnion`  constructor -- `NewThreadEvent(self.post)` -- was a duplicate of the first -- `self` -- so I removed it.

File: kitsune/wiki/models.py [hunk 2]
```diff
@@ -742,7 +743,7 @@ def gets_mozilla_account_cta(self):
             self.parent.slug if self.parent else self.slug
         ) in settings.MOZILLA_ACCOUNT_ARTICLES
 
-    @property
+    @cached_property
```
Comment (akatsoulas) [inline kitsune/wiki/models.py:746]: Nice! How about adding a method to do the calculation and then set that method here as a cached property? This way we can force also the evaluation if ever needed without having logic in the property. Eg

```python
def get_is_restricted(self):
     """Is this document restricted in terms of visibility?"""
        if self.parent:
            # Translations follow their parent's restrictions.
            return self.parent.is_restricted
        return self.restrict_to_groups.exists()

@ cached_property
def is_restricted(self):
    return self.get_is_restricted()
```

File: kitsune/wiki/models.py [hunk 3]
```diff
@@ -751,19 +752,28 @@ def is_restricted(self):
...
+            or (user.id in self.restrict_to_user_ids)
+        )
+
+    @cached_property
+    def restrict_to_user_ids(self):
+        """
+        Returns a set of user ids comprising all members of the groups specified in
+        "restrict_to_groups".
+        """
+        return set(
```
Comment (akatsoulas) [inline kitsune/wiki/models.py:775]: I would move this query into the `is_unrestricted_for` method. We are adding a property that's not used anywhere else. Let's move it in the above method and if we can extract it if we ever need it in another place.

**Review by akatsoulas — APPROVED** [review]
r+wc :guitar:

File: kitsune/wiki/models.py (hunk 2 above)
Comment (escattone) [reply kitsune/wiki/models.py:746]: That's a great idea, thanks!

**Pull request merged**